### PR TITLE
!B [#7852] fix for _rrule maximum recursion depth exceeded on multiprocessing usage

### PR DESCRIPTION
**Pull request by ahcub**
issue description:
https://github.com/matplotlib/matplotlib/issues/7852

### CODE REVIEW

**Comment by ahcub** [comment]
yes, that is correct

**Comment by ahcub** [comment]
Should I add a new test for this code change?

**Comment by efiring** [comment]
Yes, a test would be good.  Maybe in test_pickle.py?

**Comment by ahcub** [comment]
I struggle a bit with creation of test for this case without pandas
do you know how I can make it easily?
not sure that adding pandas dependency is a correct approach here...

**Comment by NelleV** [comment]
Can you create a mock object whose API is close to the one from pandas?

(pandas should not be added as a dependency).

**Comment by efiring** [comment]
I don't think anything should depend on Pandas, or needs to here.  If the problem is pickling an rrulelocator, it should be possible to test it very directly by analogy with some other tests in the test_pickle.py module.

**Comment by dopplershift** [comment]
I agree--mocking Pandas seems a bit heavy-handed, if not wrong. If the problem is pickling, test pickling.

**Comment by tacaswell** [comment]
We have some tests that run with pandas, you can just use it.  Follow the pattern from https://github.com/matplotlib/matplotlib/blob/7555a7a449b925a97b119143244a6c1caec85ceb/lib/matplotlib/tests/test_colors.py#L608

**Comment by QuLogic** [comment]
You could, but it's not really necessary to use `pandas` to reproduce the problem. Something like this is enough to trigger the recursion:
```python
>>> import matplotlib.dates
>>> import pickle
>>> r = matplotlib.dates.rrulewrapper(2)
>>> r
<matplotlib.dates.rrulewrapper object at 0x7f3c9de18470>
>>> pickle.loads(pickle.dumps(r))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
  File "/home/elliott/code/matplotlib/lib/matplotlib/dates.py", line 716, in __getattr__
    return getattr(self._rrule, name)
RecursionError: maximum recursion depth exceeded while calling a Python object
```
It just needs to be wrapped up in the test functions and given better names and such.

**Comment by ahcub** [comment]
awesome, QuLogic, thank you for the example!
I will add a test now

**Comment by ahcub** [comment]
added the test
https://github.com/matplotlib/matplotlib/pull/7854/commits/4e08061f639be62795219248c231bbb603a39415

**Comment by ahcub** [comment]
does anybody know how a new test could lead to failure of those checks?
it's a bit strange that code coverage decreased..

**Comment by jenshnielsen** [comment]
Part of the failure is our PEP8 checker, You have a blank line containing whitespaces. I suggest fixing that and let the tests rerun. I think the other failure is a random fluke

File: lib/matplotlib/tests/test_pickle.py [hunk 1]
```diff
@@ -270,6 +271,15 @@ def test_transform():
...
 
 
+def test_rrulewrapper():
+    r = rrulewrapper(2)
+    try:
+        pickle.loads(pickle.dumps(r))
+    except RecursionError:
+        print('rrulewrapper pickling test failed')
+        raise
+
```
Comment (NelleV) [inline lib/matplotlib/tests/test_pickle.py]: You still have a white space issue a this line.

Comment (ahcub) [inline lib/matplotlib/tests/test_pickle.py]: sorry, I didn't notice that
committed a change

**Comment by NelleV** [comment]
Thanks for the patch!

**Comment by tacaswell** [comment]
:+1: to backporting, but it needs a bit more work as this 'fixes' the problem via 

```
In [27]: r.__getstate__
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-27-9cc7b4e1bb10> in <module>()
----> 1 r.__getstate__

/home/tcaswell/src/p/matplotlib/lib/matplotlib/dates.py in __getattr__(self, name)
    713         self._construct.update(kwargs)
    714         self._rrule = rrule(**self._construct)
--> 715 
    716     def __getattr__(self, name):
    717         if name in self.__dict__:

AttributeError: type object 'object' has no attribute '__getattr__'

```

**Pull request merged**